### DISCUSSION TITLE
AIR 50.2.1 - IDEA Debugger not working

### DISCUSSION THREAD

**Discussion by teotigraphix** [airsdk/Adobe-Runtime-Support#2491] [Q&A]
Just asking here maybe it has come up before:

![image](https://user-images.githubusercontent.com/226417/221582644-8a90fbd4-2fbd-4e80-a738-71576285f157.png)

This just happened yesterday for the first time, I started using 50.2.1 yesterday as well.  Trying to debug on device, debugging on PC works fine I have tested that now.

The Console:

```
"C:\Program Files\JetBrains\IntelliJ IDEA 2022.2.2\jbr\bin\java.exe" -Dapplication.home=E:\Dev\AIR\SDK\AIR-SDK-Windows-50.2.1 -Dfile.encoding=UTF-8 -Djava.awt.headless=true -Duser.language=en -Duser.region=en -Xmx512m -classpath E:/Dev/AIR/SDK/AIR-SDK-Windows-50.2.1/lib/legacy/fdb.jar flex.tools.debugger.cli.DebugCLI -p 7936
Adobe fdb (Flash Player Debugger) [build development]
Copyright (c) 2004-2007 Adobe, Inc. All rights reserved.
Trying to connect to Player
Session timed out or unexpected version of the Flash Player
```

EDIT: In Logcat I can see my `trace()` calls from AIR and NativeExtensions. There is no output in IDEA's console or breakpoints.

**Comment by ajwfrost** [comment]
Does this still happen even with a really basic application? There was a slight change in the start-up of Worker objects although the debugger was coping fine with that here (there's no 'version' output, but the copyright dates match with the latest..)

Does that message come up fairly quickly i.e. when you start the remote application? or does the remote app start up but the FDB message take a while and then appear (so, timeout vs unexpected version)?

thanks

**Reply by teotigraphix** [reply]
Heh, I will have to get back to you on this one. I had a day that melted into the abyss yesterday debugging this problem. haha. I eventually went back to an earlier SDK. I need to do a release but after that I will try this again in the next day or so.

From my debugging I am really thinking it had to do with IDEA's cache and switching SDKs. I have a bunch of modules and framework libraries, it's not trivial(probably why it happened).

Was kinda funny, IDEA couldn't find Vector and I was like, what? ;-)

Will update when I switch back to 50.2.1

**Reply by ajwfrost** [reply]
> IDEA couldn't find Vector

... so back to AIR 1.1 ?! :-) 
There are some weird things sometimes with libraries/dependencies... let us know if you see this again, we did have FDB building a while ago I think so that we could debug a related issue.

thanks

**Reply by 2jfw** [reply]
> From my debugging I am really thinking it had to do with IDEA's cache and switching SDKs. I have a bunch of modules and framework libraries, it's not trivial(probably why it happened).

Rings a bell here - I have had a caching issue recently, too, with IntelliJ IDEA. Not sure how exactly I caused it but I think it was something like switching git branches either while an operation (compiling or a git commit/push) was ongoing (not even sure why I did that since it already smells fishy...).
After that, several class imports could no longer be resolved - solved it by using "Invalidate Caches..." and a restart of IDEA

**Reply by teotigraphix** [reply]
@ajwfrost Not that far! but back to `AIR-SDK-Windows-50.0.1.1` which is what I was deving with the last 2 months. I bet it was me, but still cool to make sure it wasn't. If anything else comes up I will let you know.

@2jfw Yeah I have had them too, but all in all, I love that IDE. Still, if things pick up for me, I might take a look at that plugin. ;-)

**Reply by 2jfw** [reply]
As I read `Session timed out or unexpected version of the Flash Player` this issue could be related to my post here: https://github.com/airsdk/Adobe-Runtime-Support/discussions/612
I know its not exactly the same wording but I think this can happen when the port for the connection to the player is opened and "unexpectedly" closed (e.g. hitting "compile" too quickly again or cancelling/stopping a debug session right after it launched).
If that happens, only a PC restart will help. Never was able to free the port manually or see if it is still open/occupied (even with the windows cmd tools - so my observation here may be faulty and just a good guess)

**Reply by teotigraphix** [reply]
Yeah, that does feel familiar. I wish I would have restarted my computer before I reverted SDKs. I could have answered this one right here, but then I probably wouldn't have raised the question(if it would have freed the port).

**Reply by 2jfw** [reply]
> @2jfw Yeah I have had them too, but all in all, I love that IDE. Still, if things pick up for me, I might take a look at that plugin. ;-)

I had this like two times now and the unresolvable imports where always different classes and only a very few classes where affected with  the same specific unresolvable class. However, I could find and open the class in IntelliJ - I would think the index cache might have become corrupted/flawed

**Reply by teotigraphix** [reply]
Same, I found one issue I can reproduce. For my audio stuff, I have at least 8 libraries with ANEs and different frameworks ui/uadio/app level. If I compile one in another window of IDEA, and accidentally switch to the app project that uses those libraries as a dep while the library is compiling, it will turn code red where the files are open in the app project. Closing all projects and restarting always does the trick. But it will never right itself without restarting the IDE, even though you can compile, over and over again.

**Comment by teotigraphix** [comment]
@ajwfrost Update here on the attempt at SDK 50.2.1, there is something fishy going on but I can't tell what. I have tried the most bare bones Flash app to Android and I cannot get a break point to hit in IDEA.

The console does not trace() either and all it outputs is `Trying to connect to Player` as the app runs on Android launched from IDEA.

If you wait about 2 minutes you eventually get the popup in the original post.

`Session timed out or unexpected version of the Flash Player`

So I would say something changed, I can go back SDKs and use the debugger, go forward to 50.2.1 and no debugger.

![image](https://user-images.githubusercontent.com/226417/222244657-5ac99dc2-bc6d-45b2-bfab-80fef7abe3be.png)

**Comment by ajwfrost** [comment]
We're not able to reproduce this I'm afraid, it's working fine with 50.2.1.1 here... 
I'm wondering if there's an issue with the ADB port forwarding or similar.. presumably the application just continues to run and displays the text field, rather than hanging there waiting for a debug 'continue' command? or does it stick at a black (or white) screen waiting for the start-up?

![image](https://user-images.githubusercontent.com/6400804/222770694-31f16866-a7a8-4995-b750-d7e117e2929a.png)

**Reply by teotigraphix** [reply]
The app starts up and runs as it's not being debugged, so there is no blocking, just a normal startup.

I guess I will see how things turn out then, I will deal, sometimes PCs have a mind of their own and I have a lot of things going, will update if anything changes.

Thanks.

**Reply by 2jfw** [reply]
I regularly encounter something similar with IntelliJ and something I call "Ghost Breakpoint" (see https://github.com/airsdk/Adobe-Runtime-Support/discussions/1693) - it could be related to your issue.

Could you check if your app is actually stopping at an imaginary breakpoint at the start? The app will actually _not_ stop processing and run "normally" but IntelliJ will show that it stopped at a breakpoint and it will _not_ output traces. See IntelliJ "Debugger" and check if the "Resume Program" button can be clicked - if you do not click resume quickly enough (time frame like roughly ~10s) after the issue occurs, the session will not recover and nothing gets traced and also the app may behave "strangely" and not perform as expected, kinda random not functioning, hard to describe.

This is an IntelliJ issue - here the ticket I once opened but without any comment from IntelliJ yet: [youtrack.jetbrains.com/issue/IDEA-289604](https://youtrack.jetbrains.com/issue/IDEA-289604)

**Reply by teotigraphix** [reply]
I haven't had these issues, the only issue was not hitting breakpoints (I described in the previous issue) and this issue with the debugger not connecting.

I will update the thread if I figure out what is actually causing the bad debugger connection.
